File: integration/tests/l7/cgi-bin/sample_bin.c
/* integration/tests/l7/cgi-bin/sample_bin.c */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int main() {
    // 1. Read Environment Variables
    char *method = getenv("REQUEST_METHOD");
    char *content_length = getenv("CONTENT_LENGTH");
    char *query = getenv("QUERY_STRING");

    // 2. Read Body (if any)
    char body[4096] = {0};
    int len = 0;
    if (content_length != NULL) {
        len = atoi(content_length);
        if (len > 0) {
            if (len > sizeof(body) - 1) len = sizeof(body) - 1;
            fread(body, 1, len, stdin);
        }
    }

    // 3. Output Headers
    // Must end with \r\n\r\n
    printf("Status: 200 OK\r\n");
    printf("Content-Type: text/plain\r\n");
    printf("X-CGI-Test: Vane-C-Bin\r\n");
    printf("\r\n");

    // 4. Output Body
    printf("CGI Output:\n");
    printf("Method: %s\n", method ? method : "(null)");
    printf("Query: %s\n", query ? query : "(null)");
    printf("Body Len: %d\n", len);
    printf("Body Content: %s\n", body);

    return 0;
}

File: integration/tests/l7/cgi-bin/sample_bin.lua
-- integration/tests/l7/cgi-bin/sample_bin.lua

-- Read Environment Variables
local method = os.getenv "REQUEST_METHOD" or "(null)"
local content_length = os.getenv "CONTENT_LENGTH"
local query = os.getenv "QUERY_STRING" or "(null)"

-- Read Body
local body = ""
if content_length then
	local len = tonumber(content_length)
	if len and len > 0 then
		body = io.read(len) or ""
	end
end

-- Output Headers
print "Status: 200 OK"
print "Content-Type: text/plain"
print "X-CGI-Test: Vane-Lua-Script"
print "" -- End of Headers

-- Output Body
print "CGI Output (Lua):"
print("Method: " .. method)
print("Query: " .. query)
print("Body Len: " .. (content_length or "0"))
print("Body Content: " .. body)
